clients/src/main/java/org/apache/kafka/clients/Metadata.java

```diff
@@ -146,17 +148,31 @@ public synchronized boolean updateRequested() {
     }
 
     /**
+     * If any non-retriable authentication exceptions were encountered during
+     * metadata update, clear and throw the exception.
+     */
+    public synchronized void maybeThrowAuthenticationException() {
+        if (authenticationException != null) {
+            AuthenticationException exception = authenticationException;
+            authenticationException = null;
+            throw exception;
+        }
+    }
+
+    /**
      * Wait for metadata update until the current version is larger than the last version we know of
      */
     public synchronized void awaitUpdate(final int lastVersion, final long maxWaitMs) throws InterruptedException {
         if (maxWaitMs < 0) {
-            throw new IllegalArgumentException("Max time to wait for metadata updates should not be < 0 milli seconds");
+            throw new IllegalArgumentException("Max time to wait for metadata updates should not be < 0 milliseconds");
         }
         long begin = System.currentTimeMillis();
         long remainingWaitMs = maxWaitMs;
         while (this.version <= lastVersion) {
-            if (remainingWaitMs != 0)
+            if (remainingWaitMs != 0) {
                 wait(remainingWaitMs);
+                maybeThrowAuthenticationException();
```

Comment:
Might not a big problem, but I wonder if we should check for the authentication exception before the `wait` as well? It is possible that `awaitUpdate` returns before the authentication failure happens. A subsequent call may then begin with the `authenticationException` not null which would cause a needless `wait`. I think an easy solution is to move this line up to the beginning of the `while` block.

Reply:
Thanks for the suggestion. I moved it to the top of the `while` block, outside the `if` block to give preference to authentication exception over timeout exception in the edge case.